internal/scheduler/queue/queue_cache.go

```diff
@@ -72,20 +73,44 @@ func (c *ApiQueueCache) fetchQueues(ctx *armadacontext.Context) error {
 		return err
 	}
 	queues := make([]*api.Queue, 0)
-	for {
-		msg, err := stream.Recv()
-		if err != nil {
-			return err
+	shouldEndStream := false
+
+	for !shouldEndStream {
+		msg, recvErr := stream.Recv()
+		if recvErr != nil {
+			return recvErr
 		}
 		switch msg.GetEvent().(type) {
 		case *api.StreamingQueueMessage_Queue:
 			queues = append(queues, msg.GetQueue())
 		case *api.StreamingQueueMessage_End:
 			c.queues.Store(&queues)
 			ctx.Infof("Refreshed Queues in %s", time.Since(start))
-			return nil
+			shouldEndStream = true
 		default:
 			return fmt.Errorf("unknown event of type %T", msg.GetEvent())
 		}
 	}
+
+	if closeErr := closeStream(stream); closeErr != nil {
+		return fmt.Errorf("closing get queues stream: %w", closeErr)
+	}
+
+	return nil
+}
+
+// This should be called after our end of stream message has been seen (StreamingQueueMessage_End)
+// We call recv one more time and expect an EOF back, indicating the stream is properly closed
+func closeStream(stream api.Submit_GetQueuesClient) error {
```

Comment:
where is the stream actually being closed?

Reply:
The stream is closed by the server side but the client side doesn't consider it closed until it reads the `io.EOF`. Handling `io.EOF` is well documented and the gRPC folks discuss it here: https://groups.google.com/g/grpc-io/c/XcN4hA9HonI?pli=1

The code is a bit confusing because Armada added in an additional marker to close the stream at the application layer it looks like the stream is closed but it is not until you handle `io.EOF`.